### DISCUSSION TITLE
Question - how to stream shell pipes into helix?

### DISCUSSION THREAD

**Discussion by varon** [helix-editor/helix#6379] [Q&A]
Is there any way to stream shell output instead of buffering the entire thing?
Say I wanted to run the command:
`| tail -f /var/syslog` - it won't terminate.

Is there a mechanism to do this via streaming until user cancels, adding the output as it arrives?

**Comment by the-mikedavis** [comment]
It isn't possible to interrupt the pipe for `| tail` currently. IMO streaming functionality as you say is a pager's responsibility rather than an editor. If we implement a way to cancel pipes, I think that the pipe should be like a transaction and no text should be inserted.

**Comment by varon** [comment]
Thank you for starting this discussion.

There are a number of use cases where we simply can't approach things transactionally, or rather it serves to be highly advantageous to stream information as it arrives.

Imagine a long-running computation that is gradually returning results. Those results are incrementally useful, and the decision to continue the computation depends on the results themselves.

Perhaps a more topical example would be streaming in text via one of the OpenAI APIs such as GPT. Without streaming the interaction breaks down horribly to a very boring, flow-breaking waiting process.

**Comment by varon** [comment]
This might also help with #1927

**Comment by davidzech27** [comment]
This is still really important for me in 2025. This plus command line expansions could enable streaming completions from LLMs even without plugins.

**Reply by So1aric** [reply]
I made a [draft](https://github.com/helix-editor/helix/pull/13834) few months ago, it might be broken now. This isn't enough for supporting LLMs though.

**Reply by davidzech27** [reply]
You could set up a keyboard shortcut to pass the selected text to an executable that uses an LLM to transform your code. And with command expansions you could pass it the context around the selection by passing the file name. With streaming the UX would be a lot better. This would be a sufficient LLM integration for me.
